### PR TITLE
Konstantin: Added the implementation of isPalindrom for future use cases

### PR DESCRIPTION
**Pull request by Enfree**
(cherry picked from commit 9089e7360efe04c66ae0cf262059e4ee31f6af44)

### CODE REVIEW

File: Source/com/drew/lang/StringUtil.java [hunk 1]
```diff
@@ -33,17 +33,35 @@
 /**
  * @author Drew Noakes https://drewnoakes.com
  */
-public final class StringUtil
-{
+public final class StringUtil {
+
+    @NotNull
```
Comment (ratonalgaze) [inline Source/com/drew/lang/StringUtil.java:38]: Pretending it's in our project: "we don't use this NotNull, we use that NotNull!"

Comment (ratonalgaze) [reply Source/com/drew/lang/StringUtil.java:38]: boolean can't be null

File: Source/com/drew/lang/StringUtil.java [hunk 2]
```diff
@@ -33,17 +33,35 @@
...
+    public static boolean isPalindrome(String input) {
+        String s = input;
+        if (s.length() <= 1 && s == null) {
```
Comment (ratonalgaze) [inline Source/com/drew/lang/StringUtil.java:41]: Will throw NPE. Also Null check doesn't mean it's a palindrome

File: Source/com/drew/lang/StringUtil.java [hunk 3]
```diff
@@ -33,17 +33,35 @@
...
+        }
+
+        int i = 0, j = s.length() - 1;
+        while (i < j) {
+            if (s.charAt(i) != s.charAt(j)) {
+                return false;
+            }
+            j--;
+            i++;
+        }
```
Comment (ratonalgaze) [inline Source/com/drew/lang/StringUtil.java:52]: As an option: can be replaced by Deque.

File: Tests/com/drew/lang/StringUtilTest.java [hunk 4]
```diff
@@ -23,16 +23,31 @@
...
 /**
  * @author Drew Noakes https://drewnoakes.com
  */
 public class StringUtilTest
 {
+
+    @Test
+    public void test_isPalindrome_returns_trueOnDate()
+    {
+        assertTrue(StringUtil.isPalindrome((LocalDate.now().getYear() - 23) + ""));
```
Comment (ratonalgaze) [inline Tests/com/drew/lang/StringUtilTest.java:43]: Doesn't the purpose of the isPalindrome and will break next year

File: Source/com/drew/lang/StringUtil.java [hunk 5]
```diff
@@ -33,17 +33,35 @@
 /**
  * @author Drew Noakes https://drewnoakes.com
  */
-public final class StringUtil
-{
+public final class StringUtil {
+
+    @NotNull
+    public static boolean isPalindrome(String input) {
```
Comment (ratonalgaze) [inline Source/com/drew/lang/StringUtil.java:39]: Missing javadoc for the method

File: Tests/com/drew/lang/StringUtilTest.java [hunk 6]
```diff
@@ -23,16 +23,31 @@
...
+    }
+    @Test
+    public void test_isPalindrome_returns_false()
+    {
+        assertFalse(StringUtil.isPalindrome("isPalindromeCat"));
+    }
+
```
Comment (ratonalgaze) [inline Tests/com/drew/lang/StringUtilTest.java:50]: Missing positive case test and also for invalid values like null

**Review by ratonalgaze — CHANGES_REQUESTED** [review]

File: Tests/com/drew/lang/StringUtilTest.java (hunk 6 above)
Comment (rafal-kasza) [reply Tests/com/drew/lang/StringUtilTest.java:50]: agree

File: Source/com/drew/lang/StringUtil.java (hunk 5 above)
Comment (rafal-kasza) [reply Source/com/drew/lang/StringUtil.java:39]: do we really want to use javadoc?

File: Tests/com/drew/lang/StringUtilTest.java (hunk 4 above)
Comment (rafal-kasza) [reply Tests/com/drew/lang/StringUtilTest.java:43]: agree

File: Source/com/drew/lang/StringUtil.java (hunk 2 above)
Comment (rafal-kasza) [reply Source/com/drew/lang/StringUtil.java:41]: agree

File: Source/com/drew/lang/StringUtil.java (hunk 1 above)
Comment (rafal-kasza) [reply Source/com/drew/lang/StringUtil.java:38]: ok

File: Source/com/drew/lang/StringUtil.java (hunk 3 above)
Comment (rafal-kasza) [reply Source/com/drew/lang/StringUtil.java:52]: can you elaborate?

File: Source/com/drew/lang/StringUtil.java (hunk 5 above)
Comment (ratonalgaze) [reply Source/com/drew/lang/StringUtil.java:39]: Basically I meant here "missing method description", but javadoc is the most convenient way to add it. Just comment with the explanation would suffice as well.
